### PR TITLE
Estimate PD resolution for focussed SNAP instrument

### PR DESCRIPTION
**Pull request by peterfpeterson**
The SNAP instrument wants to be able to estimate resolution for the focused instrument. This is done by averaging the contributions for the individual pixels together in quadrature.

**To test:**

The new tests, `test_focusSNAPByColumn()` and `test_focusSNAPByPanel()` run the full calculation for the focused data and compare to the values determined from only a single pixel at the focus position.

*There is no associated issue,* but this relates to [EWM1173](https://ornlrse.clm.ibmcloud.com/ccm/web/projects/Neutron%20Data%20Project%20(Change%20Management)#action=com.ibm.team.workitem.viewWorkItem&id=1173)

---

#### Reviewer ####

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Are the release notes saved in a separate file, using Issue or PR number for file name and in the correct location?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

#### Gatekeeper ####

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

File: Framework/Algorithms/test/EstimateResolutionDiffractionTest.h [hunk 1]
```diff
@@ -102,18 +121,116 @@ class EstimateResolutionDiffractionTest : public CxxTest::TestSuite {
...
+  /*
+   * This test is meant to duplicate a workflow where data from SNAP_57514 is time focussed before calculating the
+   * resolution of the resulting 6 spectra. The values checked against are from observations from doing individual peak
+   * fits.
+   */
+  void test_focusSNAPByColumn() {
+    auto ws = createSNAPLiteInstrument("Column");
+    const auto WS_IN = ws->getName();
+    std::string WS_OUT("SNAPColumn_Resolution");
+    constexpr double deltaTOFOverTOF{0.001 / 15};
```
Comment (mguthriem) [inline Framework/Algorithms/test/EstimateResolutionDiffractionTest.h:136]:
```suggestion
    constexpr double deltaTOFOverTOF{0.002};
```

**Comment by sf1919** [comment]
@peterfpeterson I cancelled the Windows tests as the System tests had passed and you have been asked to make changes.

File: Framework/Algorithms/src/EstimateResolutionDiffraction.cpp [hunk 2]
```diff
@@ -225,100 +226,130 @@ void EstimateResolutionDiffraction::estimateDetectorResolution() {
...
-      term1Sq = term1Sq * term1Sq;
+      const auto centreVel = m_centreVelocity;
+      const auto deltaT = m_deltaT;
+      term1Sq = std::accumulate(spectrumDefinition.cbegin(), spectrumDefinition.cend(), 0.,
+                                [&detectorInfo, &centreVel, &l1, &deltaT](const auto sum, const auto &index) {
+                                  if (detectorInfo.isMasked(index.first) || detectorInfo.isMonitor(index.first)) {
+                                    return sum;
+                                  } else {
+                                    // Get the distance from detector to source
+                                    const double l2 = detectorInfo.l2(index.first);
```
Comment (rboston628) [inline Framework/Algorithms/src/EstimateResolutionDiffraction.cpp:250]: should `detectorInfo` be `spectrumInfo`?

File: Framework/Algorithms/src/EstimateResolutionDiffraction.cpp [hunk 3]
```diff
@@ -225,100 +226,130 @@ void EstimateResolutionDiffraction::estimateDetectorResolution() {
...
+                                    const auto &detector = detectorInfo.detector(index.first);
+                                    const auto realdet = dynamic_cast<const Detector *>(&detector);
+                                    if (realdet) {
+                                      const auto l2 = detectorInfo.l2(index.first);
+                                      const auto l_total = (l1 + l2);
+                                      realdet->shape();
+                                      const double dx = realdet->getWidth();
+                                      const double dy = realdet->getHeight();
+                                      const double dz = realdet->getDepth();
+                                      const double detdimSq = (dx * dx + dy * dy + dz * dz) * (0.5 * 0.5);
```
Comment (rboston628) [inline Framework/Algorithms/src/EstimateResolutionDiffraction.cpp:277]: `0.25`?

Comment (rboston628) [reply Framework/Algorithms/src/EstimateResolutionDiffraction.cpp:277]: I'm not sure why we would want 1/4 of the square distance, either

File: Framework/Algorithms/src/EstimateResolutionDiffraction.cpp [hunk 4]
```diff
@@ -225,100 +226,130 @@ void EstimateResolutionDiffraction::estimateDetectorResolution() {
...
+      term2Sq = term2Sq / numDet;
     }
-    const double l_total = (l1 + l2);
-    const double term2Sq = (detdimSq + m_sourceDeltaLMetersSq) / (l_total * l_total);
 
-    // resolution in angle
-    const double twotheta = spectrumInfo.isMonitor(i) ? 0.0 : spectrumInfo.twoTheta(i);
-    const double theta = 0.5 * twotheta;
+    // resolution in angle - everything is in radians
+    const double theta = spectrumInfo.isMonitor(specNum) ? 0.0 : 0.5 * spectrumInfo.twoTheta(specNum);
```
Comment (rboston628) [inline Framework/Algorithms/src/EstimateResolutionDiffraction.cpp:289]: why not maintain `twotheta`?  It avoids having to write `2.0 * theta` twice below

File: Framework/Algorithms/src/EstimateResolutionDiffraction.cpp [hunk 5]
```diff
@@ -225,100 +226,130 @@ void EstimateResolutionDiffraction::estimateDetectorResolution() {
...
-                          });
-      deltathetaSq = solidangle;
+      const double sourceTerm = m_sourceDeltaThetaRadiansSq;
+      const double solidAngle = std::accumulate(
+          spectrumDefinition.cbegin(), spectrumDefinition.cend(), 0.,
+          [&componentInfo, &samplepos, &detectorInfo, &sourceTerm](const auto sum, const auto &index) {
+            if (detectorInfo.isMasked(index.first) || detectorInfo.isMonitor(index.first)) {
+              return sum;
+            } else {
+              const double theta = 0.5 * detectorInfo.twoTheta(index.first); // angle of the pixel
```
Comment (rboston628) [inline Framework/Algorithms/src/EstimateResolutionDiffraction.cpp:305]: How does this  theta differ from the theta in line 289?

Comment (peterfpeterson) [reply Framework/Algorithms/src/EstimateResolutionDiffraction.cpp:305]: This is for the pixel where that was for the average over the spectrum

**Review by rboston628 — APPROVED** [review]

**Pull request merged**